### DISCUSSION TITLE
Show execution order of scripts

### DISCUSSION THREAD

**Discussion by 2-4601** [twpayne/chezmoi#3491] [Q&A] [answered]
Let's say I have the following chezmoi scripts in my source state:
```
$ find $(chezmoi source-path) -name 'run_*' | sort
~/.local/share/chezmoi/home/.chezmoiscripts/run_onchange_after_010-mkdir-xdg-state.sh
~/.local/share/chezmoi/home/.chezmoiscripts/run_onchange_after_100-import-gnome-settings.sh.tmpl
~/.local/share/chezmoi/home/.chezmoiscripts/run_onchange_before_100-alpine-install-bash.sh.tmpl
~/.local/share/chezmoi/home/.chezmoiscripts/run_onchange_before_200-install-system-packages.sh.tmpl
~/.local/share/chezmoi/home/.chezmoiscripts/run_onchange_before_300-install-foreign-packages.sh.tmpl
~/.local/share/chezmoi/home/dot_config/fish/fisher/run_onchange_before_update-fisher.fish.tmpl
~/.local/share/chezmoi/home/dot_local/share/applications/run_onchange_after_update-desktop-database.sh.tmpl
```
If I want chezmoi to list all the scripts, I can run:
```
$ chezmoi managed --include scripts
.chezmoiscripts/010-mkdir-xdg-state.sh
.chezmoiscripts/100-alpine-install-bash.sh
.chezmoiscripts/100-import-gnome-settings.sh
.chezmoiscripts/200-install-system-packages.sh
.chezmoiscripts/300-install-foreign-packages.sh
.config/fish/fisher/update-fisher.fish
.local/share/applications/update-desktop-database.sh
```
Is there a way to make chezmoi show the order in which they would be executed (e.g., `before` scripts would show before `after` scripts)?
In short, I'm looking for something like this:
```
$ chezmoi <show-scripts-in-execution-order>
.chezmoiscripts/100-alpine-install-bash.sh
.chezmoiscripts/200-install-system-packages.sh
.chezmoiscripts/300-install-foreign-packages.sh
.config/fish/fisher/update-fisher.fish
.chezmoiscripts/010-mkdir-xdg-state.sh
.chezmoiscripts/100-import-gnome-settings.sh
.local/share/applications/update-desktop-database.sh
```

**Comment by twpayne** [comment]
Yes. You can run `chezmoi status --include=scripts`. Note that this will show you the scripts that `chezmoi apply` and will not include `run_once_` and `run_onchange_` scripts if they do not need to be run.

**Reply by 2-4601** [reply]
Thank you.

Is it possible to temporarily ignore the current state of my scripts? (All of my scripts do not need to be run currently, but it would be helpful to still see the order.)

I tried to ignore the state with:
```
$ chezmoi status --include scripts --persistent-state emptyfile
chezmoi: open ~/emptyfile: write ~/emptyfile: bad file descriptor
```

**Edit:**
I was able to create an empty `chezmoistate.boltdb` by initializing another chezmoi repository and the using that file as a persistent state for my real chezmoi repository. This gives me the list of scripts in the right order.

```
$ chezmoi status --include scripts --persistent-state dummy_chezmoistate.boltdb
R .chezmoiscripts/100-alpine-install-bash.sh
R .chezmoiscripts/200-install-system-packages.sh
R .chezmoiscripts/300-install-foreign-packages.sh
R .config/fish/fisher/update-fisher.fish
R .chezmoiscripts/010-mkdir-xdg-state.sh
R .chezmoiscripts/100-import-gnome-settings.sh
R .local/share/applications/update-desktop-database.sh
```

**Reply by Lockszmith-GH** [reply]
~try `/dev/null`~ See my updated response below

**Reply by 2-4601** [reply]
That was one of the first things I tried :)
`/dev/null`, `<(echo -n)` all fail in various ways. Only a dummy bolt DB file works.

**Reply by Lockszmith-GH** [reply]
bummer, ~I wonder if~ there is a way other than tricking chezmoi though.

Edit: yes, there is [`managed`](https://www.chezmoi.io/reference/commands/managed/):
```shell
chezmoi managed --include scripts
```

**Reply by 2-4601** [reply]
Appreciate the help. The `managed` command indeed lists the scripts (I also quickly mentioned this in the original post) but it lists them in an alphabetical order. I am looking for a list that is sorted by the order of execution. Motivation is that since there could be several scripts in different places, it would help to see that they are run in the order I intended.

The `status` command with an empty bolt DB does the trick. It's a bit of a hack, but I'm happy that it can be achieved with existing functionality. Thanks for the help and suggestions.

**Reply by 2-4601** [reply]
:exclamation: **Edit:** _This is an unnecessary complex solution. Please see the marked answer below for a proper solution_ :exclamation: 

1. Create a new "empty" `chezmoistate.boltdb` file (read the full thread to see why it is needed):
```
$ env --ignore-environment bash -c 'export HOME=/tmp && cd $HOME && touch emptyfile && chezmoi init && chezmoi add emptyfile && chezmoi apply'
```

2. List the status of your scripts (by comparing to the empty bolt database state):
```
$ chezmoi status --include scripts --persistent-state /tmp/.config/chezmoi/chezmoistate.boltdb

chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file
R .chezmoiscripts/100-alpine-install-bash.sh
R .chezmoiscripts/200-install-system-packages.sh
R .chezmoiscripts/300-install-foreign-packages.sh
R .config/fish/fisher/update-fisher.fish
R .chezmoiscripts/010-mkdir-xdg-state.sh
R .chezmoiscripts/100-import-gnome-settings.sh
R .local/share/applications/update-desktop-database.sh
```

**Reply by Lockszmith-GH** [reply]
> Appreciate the help. ...(I also quickly mentioned this in the original post)

Apologies - just my swiss cheese brain, sorry.

**Comment by twpayne** [comment, marked as answer]
Thanks for following up here - this is really helpful to people who have this problem in the future!

> 1. Create a new "empty" `chezmoistate.boltdb` file (read the full thread to see why it is needed):

This is not needed. Using a *non-existent* state file is sufficient. chezmoi will create the state file. The problem with using an empty state file is that an empty file is not a valid BoltDB database.

You can use:

```console
$ chezmoi status --persistent-state=$(mktemp -d)/chezmoistate.boltdb --include=scripts
```

**Reply by 2-4601** [reply]
Oh wow, that's so simple! I don't know why my brain insisted on trying empty files that _exist_! :sweat_smile:
